storm-server/src/main/java/org/apache/storm/metric/StormMetricsRegistry.java

```diff
@@ -27,52 +28,63 @@
 
 @SuppressWarnings("unchecked")
 public class StormMetricsRegistry {
-    public static final MetricRegistry DEFAULT_REGISTRY = new MetricRegistry();
+    private static final MetricRegistry DEFAULT_REGISTRY = new MetricRegistry();
     private static final Logger LOG = LoggerFactory.getLogger(StormMetricsRegistry.class);
 
-    public static Meter registerMeter(String name) {
-        Meter meter = new Meter();
-        return register(name, meter);
+    public static Meter registerMeter(final String name) {
+        return register(name, new Meter());
     }
 
-    // TODO: should replace Callable to Gauge<Integer> when nimbus.clj is translated to java
-    public static Gauge<Integer> registerGauge(final String name, final Callable fn) {
-        Gauge<Integer> gauge = new Gauge<Integer>() {
-            @Override
-            public Integer getValue() {
-                try {
-                    return (Integer) fn.call();
-                } catch (Exception e) {
-                    LOG.error("Error getting gauge value for {}", name, e);
-                }
-                return 0;
+    /**
+     * Register a gauge with provided callback.
+     * @param name name of the gauge
+     * @param fn callback that measures
+     * @param <V> type of measurement the callback returns, also the type of gauge
+     * @return registered gauge
+     */
+    public static <V> Gauge<V> registerGauge(final String name, final Callable<V> fn) {
+        return register(name, () -> {
+            try {
+                return fn.call();
+            } catch (Exception e) {
+                LOG.error("Error getting gauge value for {}", name, e);
             }
-        };
-        return register(name, gauge);
+            return null;
+        });
     }
 
-    public static void registerProvidedGauge(final String name, Gauge gauge) {
+    /**
+     * Register a provided gauge. Use this method if custom gauges is needed or
+     * no checked exceptions should be handled.
+     * @param name name of the gauge
+     * @param gauge gauge
+     * @param <V> type of value the gauge measures
+     */
+    public static <V> void registerProvidedGauge(final String name, final Gauge<V> gauge) {
         register(name, gauge);
     }
 
     public static Histogram registerHistogram(String name, Reservoir reservoir) {
-        Histogram histogram = new Histogram(reservoir);
-        return register(name, histogram);
+        return register(name, new Histogram(reservoir));
+    }
+
+    public static void registerAll(final String prefix, MetricSet metrics) {
```

Comment:
Why is this method necessary?

Reply:
This is used in #2764. Should move the change there?

Reply:
Having it here is fine, but I'm not sure I understand what it does differently from just calling the register method directly? Is it just for clarity?

Reply:
For wrapper to the private `register()` method. I don't see any problem if we should make `register` public though.

Reply:
I think it's a little weird to have this and `unregister` be public, but not the generic `register` method.

Reply:
I don't have a strong opinion on this, but making the `register` method public might make for a nicer API, especially since `unregister` can unregister any kind of metric.

Reply:
I'll have to think about this a bit more, since I will have a few more PR depending on this. The way `MetricRegistry` handles `MetricSet` under the hood is a bit wacky, and they haven't supported `removeAll(MetricSet set)` method in their release yet. Do you have any advice of improving `StormMetricsRegistry` in general?

Reply:
Improved in #2771

Reply:
Okay, should I first merge change in this PR into #2771 and then port them to where they are actually invoked?